File: netlogo-gui/src/main/prim/_observervariable.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Observer;
import org.nlogo.core.Syntax;
import org.nlogo.core.Reference;
import org.nlogo.core.AgentKindJ;
import org.nlogo.nvm.Reporter;

public final strictfp class _observervariable
    extends Reporter {
  public int vn = 0;

  public _observervariable(int vn) {
    this.vn = vn;
  }

  public Reference makeReference() {
    return new Reference(AgentKindJ.Observer(), vn, this);
  }

  @Override
  public Object report(final org.nlogo.nvm.Context context) {
    return world.observer().getVariable(vn);
  }

  @Override
  public String toString() {
    if (world != null) {
      return super.toString() + ":" + world.observerOwnsNameAt(vn);
    } else {
      return super.toString() + ":" + vn;
    }
  }

  public Object report_1(final org.nlogo.nvm.Context context) {
    return world.observer().getVariable(vn);
  }
}

File: netlogo-headless/src/main/prim/_observervariable.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.{ AgentKind, Reference, Referenceable }
import org.nlogo.nvm.{ Context, Reporter }

class _observervariable(_vn: Int) extends Reporter with Referenceable {

  override def toString =
    super.toString + ":" +
      Option(world).map(_.observerOwnsNameAt(vn)).getOrElse(vn.toString)

  // MethodRipper won't let us call a public method from report_1()
  // so we must keep vn and _vn separate - ST 9/22/12
  def vn = _vn

  def makeReference =
    new Reference(AgentKind.Observer, _vn, this)

  override def report(context: Context): AnyRef =
    world.observer.getVariable(_vn)

  def report_1(context: Context): AnyRef =
    world.observer.getVariable(_vn)

}
